packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -136,8 +136,10 @@ class RenderParagraph extends RenderBox
     assert(value != null);
     switch (_textPainter.text!.compareTo(value)) {
       case RenderComparison.identical:
-      case RenderComparison.metadata:
         return;
+      case RenderComparison.metadata:
+        _textPainter.text = value;
```

Comment:
Okay. I think `_cachedAttributedLabel` should not be included also. There is no connection between this property and `metadata`.

```dart
...
      case RenderComparison.metadata:
        _textPainter.text = value;
        _cachedCombinedSemanticsInfos = null;
        markNeedsSemanticsUpdate();
        break;
      case RenderComparison.paint:
        _textPainter.text = value;
        _cachedAttributedLabel = null;
        _cachedCombinedSemanticsInfos = null;
        _extractPlaceholderSpans(value);
        markNeedsPaint();
        markNeedsSemanticsUpdate();
        break;
...
```

Reply:
@tgucio Any suggestions about the test for `SemanticsNode`? I have viewed the `paragraph_test.dart` and have no idea about what I should include.

Reply:
There's a specific test for gesture recognizers in semantics: `test('Supports gesture recognizer semantics', () {` - it seems unfinished though as it doesn't check the contents of SemanticsNode.

So perhaps you could do:
- in the existing test , add a check (I'm guessing gestures should be translated to .actions)
- in the new test, start with onTap recognizer, check the node, then change it to onLongPress and check the node again

Reply:
@tgucio  After a series of tests, I found that even if `markNeedsSemanticsUpdate` is executed, the `SemanticsNode` will not be updated, and even if `pumpFrame(phase: EnginePhase.layout);` is executed, the `SemanticsNode` will not be updated also. So I think this is another historical problem. Here's my test code.

```dart
  test('can just update the gesture recognizer', () async {
    final TapGestureRecognizer recognizerBefore = TapGestureRecognizer()..onTap = () {};
    final RenderParagraph paragraph = RenderParagraph(
      TextSpan(text: 'How are you \n', recognizer: recognizerBefore),
      textDirection: TextDirection.ltr,
    );
    layout(paragraph);
    expect((paragraph.text as TextSpan).recognizer, same(recognizerBefore));
    final SemanticsNode nodeBefore = SemanticsNode();
    paragraph.assembleSemanticsNode(nodeBefore, SemanticsConfiguration(), <SemanticsNode>[]);
    nodeBefore.visitChildren((SemanticsNode child) {
      final SemanticsData data = child.getSemanticsData();
      expect(data.hasAction(SemanticsAction.tap), true);
      return true;
    });
    final LongPressGestureRecognizer recognizerAfter = LongPressGestureRecognizer()..onLongPress = () {};
    paragraph.text = TextSpan(text: 'How are you \n', recognizer: recognizerAfter);
    pumpFrame(phase: EnginePhase.paint);
    expect((paragraph.text as TextSpan).recognizer, same(recognizerAfter));
    final SemanticsNode nodeAfter = SemanticsNode();
    paragraph.assembleSemanticsNode(nodeAfter, SemanticsConfiguration(), <SemanticsNode>[]);
    nodeAfter.visitChildren((SemanticsNode child) {
      final SemanticsData data = child.getSemanticsData();
      // An error will be reported here!
      expect(data.hasAction(SemanticsAction.longPress), true);
      expect(data.hasAction(SemanticsAction.tap), false);
      return true;
    });
  });
```

Considering that this PR is to fix the gesture recognizer issue, I suggest focusing on the part of gesture recognizer first and then opening a new issue to fix the issue about the `SemanticsNode`.

Reply:
When you call `layout()` in tests, the default phase is `EnginePhase.layout` which won't update semantics I believe. So you'll probably need to specify this:
```
layout(tree, phase: EnginePhase.flushSemantics);
```

For subsequent updates it should be enough to pump a frame with the same phase:
```
pumpFrame(phase: EnginePhase.flushSemantics);
```

See `test/rendering/simple_semantics_test.dart` for reference, `test('only send semantics update if semantics have changed'`.

Could you give this a try? If it gets too complex, we can separate into another PR.

Reply:
It works! I checked the code, and it is true that the owner is required to ensure that `pumpFrame(phase: EnginePhase.flushSemantics)` executes successfully. Otherwise, the following function will return directly at the beginning.

<details>
  <summary>ScreenShot</summary>
 
<img width="1624" alt="Screenshot 2022-12-16 at 09 32 19" src="https://user-images.githubusercontent.com/11863087/208002132-09009e75-d26e-43b4-9d20-6d0c1a98b5d5.png">

</details>

Reply:
Actually I think this was coming from another place. `markNeedsSemanticsUpdate()` will just set the cached value to null which is OK. But when semantics are flushed, nothing happens when the node has no owner, see `object.dart`:

```
  void flushSemantics() {
    if (_semanticsOwner == null) {
      return;
    }
```